### PR TITLE
Fix sending irrelevant block updates  & Fix NullPointerException exploit

### PR DESCRIPTION
**Pull request by crafter23456**
[FlamePaper-0106] Fix sending irrelevant block updates to the client
[FlamePaper-0110] Fix NullPointerException exploits for invalid logins

### CODE REVIEW

File: NachoSpigot-Server/src/main/java/net/minecraft/server/World.java [hunk 1]
```diff
@@ -497,7 +497,11 @@ public boolean setTypeAndData(BlockPosition blockposition, IBlockData iblockdata
 
     // CraftBukkit start - Split off from original setTypeAndData(int i, int j, int k, Block block, int l, int i1) method in order to directly send client and physic updates
     public void notifyAndUpdatePhysics(BlockPosition blockposition, Chunk chunk, Block oldBlock, Block newBLock, int flag) {
-        if ((flag & 2) != 0 && (chunk == null || chunk.isReady())) {  // allow chunk to be null here as chunk.isReady() is false when we send our notification during block placement
+        if ((flag & 2) != 0 && (!this.isClientSide || (flag & 4) == 0) && (chunk == null || chunk.isReady())) {  // allow chunk to be null here as chunk.isReady() is false when we send our notification during block placement
```
Comment (Lucaskyy) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/World.java:500]: This doesn't compile, here is the fix:
```
    public void notifyAndUpdatePhysics(BlockPosition blockposition, Chunk chunk, Block oldBlock, Block newBLock, int flag) {
        if ((flag & 2) != 0 && (!this.isClientSide || (flag & 4) == 0) && (chunk == null || chunk.isReady())) {  // allow chunk to be null here as chunk.isReady() is false when we send our notification during block placement
            this.notify(blockposition);
        }

        if (!this.isClientSide && (flag & 1) != 0) {
            this.update(blockposition, oldBlock);
            if (newBLock.isComplexRedstone()) {
                this.updateAdjacentComparators(blockposition, newBLock);
            }
        }
    }
```

**Review by Lucaskyy — CHANGES_REQUESTED** [review]
Please fix this issue :)

File: NachoSpigot-Server/src/main/java/net/minecraft/server/World.java (hunk 1 above)
Comment (crafter23456) [reply NachoSpigot-Server/src/main/java/net/minecraft/server/World.java:500]: Fixed https://github.com/CobbleSword/NachoSpigot/pull/109/commits/37d9f48e3a6136348b9f56e75444194f512004d3

**Comment by Lucaskyy** [comment]
Thank you :)

**Pull request merged**